core/sdk-core/src/test/java/software/amazon/awssdk/core/internal/util/MetricUtilsTest.java

```diff
@@ -0,0 +1,144 @@
+import software.amazon.awssdk.metrics.MetricPublisher;
+import software.amazon.awssdk.utils.Pair;
+
+public class MetricUtilsTest {
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void testMeasureDuration_returnsAccurateDurationInformation() {
+        long testDurationNanos = Duration.ofMillis(1).toNanos();
+
+        Pair<Object, Duration> measuredExecute = MetricUtils.measureDuration(() -> {
+            long start = System.nanoTime();
+            // spin thread instead of Thread.sleep() for a bit more accuracy...
+            while (System.nanoTime() - start < testDurationNanos) {
+            }
+            return "foo";
+        });
+
+        assertThat(measuredExecute.right()).isGreaterThanOrEqualTo(Duration.ofNanos(testDurationNanos));
+    }
+
+    @Test
+    public void testMeasureDuration_returnsCallableReturnValue() {
+        String result = "foo";
+
+        Pair<String, Duration> measuredExecute = MetricUtils.measureDuration(() -> result);
+
+        assertThat(measuredExecute.left()).isEqualTo(result);
+    }
+
+    @Test
+    public void testMeasureDurationUnsafe_doesNotWrapException() throws Exception {
+        IOException ioe = new IOException("boom");
+
+        thrown.expect(IOException.class);
+        try {
+            MetricUtils.measureDurationUnsafe(() -> {
+                throw ioe;
+            });
+        } catch (IOException caught) {
+            assertThat(caught).isSameAs(ioe);
+            throw caught;
+        }
+    }
+
+    @Test
+    public void testMeasureDuration_doesNotWrapException() {
+        RuntimeException e = new RuntimeException("boom");
+
+        thrown.expect(RuntimeException.class);
+
+        try {
+            MetricUtils.measureDuration(() -> {
+                throw e;
+            });
+        } catch (RuntimeException caught) {
+            assertThat(caught).isSameAs(e);
+            throw caught;
+        }
+    }
+
+    @Test
+    public void testCollectHttpMetrics_collectsAllExpectedMetrics() {
+        MetricCollector mockCollector = mock(MetricCollector.class);
+
+        int statusCode = 200;
+        String requestId = "request-id";
+        String amznRequestId = "amzn-request-id";
+        String requestId2 = "request-id-2";
+
+        SdkHttpFullResponse response = SdkHttpFullResponse.builder()
+                .statusCode(statusCode)
+                .putHeader("x-amz-request-id", requestId)
+                .putHeader(HttpResponseHandler.X_AMZN_REQUEST_ID_HEADER, amznRequestId)
+                .putHeader(HttpResponseHandler.X_AMZ_ID_2_HEADER, requestId2)
+                .build();
+
+        MetricUtils.collectHttpMetrics(mockCollector, response);
+
+        verify(mockCollector).reportMetric(CoreMetric.HTTP_STATUS_CODE, statusCode);
+        verify(mockCollector).reportMetric(CoreMetric.AWS_REQUEST_ID, requestId);
+        verify(mockCollector).reportMetric(CoreMetric.AWS_REQUEST_ID, amznRequestId);
```

Comment:
Why do we report the same metric with two different values?

Reply:
Depending on the service, the Header used for the request ID can be differ between these two values. This is just to ensure we're catching both cases, I don't think a service will use both.

Reply:
ok!